File: dsa/py/basics/binary_search.py
"""
Applicable only for sorted array

"""


def binary_search_recursive(arr, element, low, high):
    mid = (low + high) // 2

    if low <= high:
        # base case
        if arr[mid] == element:
            return mid

        # search right arr
        if element > arr[mid]:
            return binary_search_recursive(arr, element, mid + 1, high)

        # search left arr
        else:
            return binary_search_recursive(arr, element, low, mid - 1)

    return -1


def binary_search_iterative(arr, element):
    low = 0
    high = len(arr) - 1

    while low <= high:
        mid = (low + high) // 2
        if element == arr[mid]:
            return mid
        if element > arr[mid]:
            low = mid + 1
        else:
            high = mid - 1

    return -1


arr = [1, 2, 3, 4, 5]
element = 6

print(binary_search_recursive(arr, element, 0, len(arr) - 1))
print(binary_search_iterative(arr, element))

File: dsa/java/DSA/Concepts/Searching/BinarySearch/binary_search.java
package Algorithms.Searching.BinarySearch;

import Algorithms.Sorting.merge_sort;
import Common.io;

public class binary_search {

    // Recursive
    public static int recursive_bs(int[] arr, int key, int low, int high) { // 1,2,3,4
        int mid = (high + low) / 2;
        if (low <= high) {
            if (key == arr[mid]) {
                return mid;
            } else if (key > arr[mid]) { // 4 > 2
                return recursive_bs(arr, key, mid + 1, high);
            } else {
                return recursive_bs(arr, key, low, mid - 1);
            }
        }
        return -1;
    }

    // Iterative
    public static int iterative_bs(int[] arr, int key) {
        int low = 0;
        int high = arr.length - 1;
        int mid = -1;
        while (low <= high) {
            mid = (low + high) / 2;

            if (arr[mid] == key) {
                return mid;
            } else if (key > arr[mid]) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }
        return mid;
    }

    public static void main(String[] args) {

        int[] arr = io.readArr();
        arr = merge_sort.ms(arr, 0, arr.length - 1);
        // Arrays.sort(arr); // Input array must be sorted

        io.displayArr(arr);

        int element = io.readInt();

        // int index = recursive_bs(arr, element, 0, arr.length-1);
        int index = iterative_bs(arr, element);
        System.out.print("Index = ");
        io.displayInt(index);
    }
}
